flocker/testtools/_flaky.py

```diff
@@ -22,10 +43,274 @@ def flaky(jira_key):
+        # reach the minimum number of successes.
+        max_fails = flaky.max_runs - flaky.min_passes
+        while (successes < flaky.min_passes and
+               len(results) - successes <= max_fails):
+            was_successful, result_type, details = self._attempt_test(case)
+            if was_successful:
+                successes += 1
+            results.append((result_type, details))
+        successful = successes >= flaky.min_passes
+
+        flaky_data = flaky.to_dict()
+        flaky_data.update({'runs': len(results), 'passes': successes})
+        flaky_details = {
+            'flaky': text_content(pformat(flaky_data)),
+        }
+        combined_details = _combine_details(
+            [flaky_details] + list(r[1] for r in results))
+
+        result.startTest(case)
+        if successful:
+            skip_reported = False
+            for result_type, details in results:
+                if result_type == _ResultType.skip:
+                    result.addSkip(case, details=details)
+                    skip_reported = True
+
+            if not skip_reported:
+                self._output.write(
+                    '@flaky(%s): passed %d out of %d runs '
+                    '(min passes: %d; max runs: %d)'
+                    % (case.id(), successes, len(results), flaky.min_passes,
+                       flaky.max_runs)
+                )
+
+                result.addSuccess(case, details=combined_details)
+        else:
+            # XXX: How are we going to report on tests that sometimes fail,
+            # sometimes error, sometimes skip? Currently we just error.
+            result.addError(case, details=combined_details)
+        result.stopTest(case)
+        return result
+
+    def _attempt_test(self, case):
+        """
+        Run 'case' with a temporary result.
+
+        :param testtools.TestCase case: The test to run.
+
+        :return: a tuple of ``(successful, result, details)``, where
+            ``successful`` is a boolean indicating whether the test was
+            succcessful, ``result`` is a _ResultType indicating what the test
+            result was and ``details`` is a dictionary of testtools details.
+        """
+        tmp_result = testtools.TestResult()
+        self._run_test(case, tmp_result)
+        result_type = _get_result_type(tmp_result)
+        details = pmap(case.getDetails())
+        if result_type == _ResultType.skip:
+            # XXX: Work around a testtools bug where it reports stack traces
+            # for skips that aren't passed through its supported
+            # SkipException: https://bugs.launchpad.net/testtools/+bug/1518100
+            [reason] = list(tmp_result.skip_reasons.keys())
+            details = details.discard('traceback').set(
+                'reason', text_content(reason))
+        _reset_case(case)
+        return (tmp_result.wasSuccessful(), result_type, details)
+
+
+class _ResultType(Names):
+    """
+    Different kinds of test results.
```

Comment:
:grey_question: I feel like as a company we usually document `NamedConstant`'s. In my opinion there is no reason to do that in this case, as these are just constants to describe states that come from a dependency and are probably as self-describing as any documentation that we would add.

Just pointing it out in case you are of the opinion that we should document them and just forgot.